runtime/grid_reduction.cu

```diff
@@ -623,4 +623,31 @@ __device__ void gridReduceGroup(
 }
 #endif // NVFUSER_PROFILE_KERNEL
 
+// This performs a single reduction step, combining a single element "in" with
+// a previous value "work". If this is the first step, we ignore "work". If
+// this is not the last step, we write the result of this step back to "work".
+template <typename T, typename Func>
+__device__ void serialReductionStep(
+    T& out,
+    T in,
+    T init,
+    volatile T& work,
+    Func reduction_op,
+    bool first_step,
+    bool last_step,
+    bool read_pred,
+    bool write_pred) {
```

Comment:
Not clear how `write_pred` is supposed to be used. Should `work` be updated even when `write_pred` is false?